packages/@aws-cdk/core/lib/private/cfn-reference.ts

```diff
@@ -55,26 +55,30 @@ export class CfnReference extends Reference {
    */
   public static for(target: CfnElement, attribute: string, refRender?: ReferenceRendering) {
     return CfnReference.singletonReference(target, attribute, refRender, () => {
-      const cfnIntrinsic = refRender === ReferenceRendering.FN_SUB
-        ? ('${' + target.logicalId + (attribute === 'Ref' ? '' : `.${attribute}`) + '}')
-        : (attribute === 'Ref'
-          ? { Ref: target.logicalId }
-          : {
-            'Fn::GetAtt': refRender === ReferenceRendering.GET_ATT_STRING
-              ? `${target.logicalId}.${attribute}`
-              : [target.logicalId, attribute],
-          }
-        );
+      let cfnIntrinsic: any;
+      switch (refRender) {
+        // The first 2 are only used by cfn-parse, so it's fine if they're round-tripped exactly.
+        // They could never reference a cross-stack construct anyway.
+        case ReferenceRendering.FN_SUB:
+          cfnIntrinsic = '${' + target.logicalId + (attribute === 'Ref' ? '' : `.${attribute}`) + '}';
+          break;
+        case ReferenceRendering.GET_ATT_STRING:
+          cfnIntrinsic = { 'Fn::GetAtt': `${target.logicalId}.${attribute}` };
+          break;
+        default:
+          cfnIntrinsic = { '$Cdk::Ref': [target.node.path, attribute] };
+          break;
+      }
       return new CfnReference(cfnIntrinsic, attribute, target);
     });
   }
 
   /**
-   * Return a CfnReference that references a pseudo referencd
+   * Return a CfnReference that references a scoped pseudo
    */
   public static forPseudo(pseudoName: string, scope: Construct) {
     return CfnReference.singletonReference(scope, `Pseudo:${pseudoName}`, undefined, () => {
-      const cfnIntrinsic = { Ref: pseudoName };
+      const cfnIntrinsic = { '$Cdk::Ref': [scope.node.path, pseudoName] };
```

Comment:
Define a `const` for `"$CDK::Ref"` and use it everywhere. Also add a comment explaining what is it

```suggestion
      const cfnIntrinsic = { '$CDK::Ref': [scope.node.path, pseudoName] };
```